Sources/NIOHTTP2/HTTP2FrameParser.swift

```diff
@@ -0,0 +1,738 @@
+    ///
+    /// The method's effect on the input buffer and its return value depend on the content of
+    /// the buffer:
+    ///
+    /// - If the input buffer contains more bytes beyond the current frame, those bytes will
+    ///   remain in the buffer; you may continue calling `decode(bytes:)` until the buffer is
+    ///   empty, or until its size does not change.
+    /// - If a given buffer does not contain a complete frame header (9 bytes), this method
+    ///   will take no action, and the buffer will remain unchanged when the method returns.
+    /// - If the buffer does not contain an entire frame including payload, it will be cached
+    ///   internally by the decoder, and the next call to `decode(bytes:)` will append any
+    ///   new bytes to that cache in an effort to complete that frame's payload.
+    /// - When enough bytes to decode an entire frame have been accumulated or provided, the
+    ///   resultant frame will be returned. If the entire frame and its payload have not yet
+    ///   been accumulated, `nil` will be returned.
+    /// - If we gather enough bytes to construct a frame, but that frame is invalid or malformed,
+    ///   we consume all the bytes and throw the error specified in RFC 7540.
+    ///
+    /// - note: The HTTP/2 protocol specifies that unknown frame types should be ignored. This
+    ///         implementation will silently collect and drop all bytes associated with frames
+    ///         whose type is unrecognized; it will not throw an error or otherwise provide
+    ///         those bytes for processing elsewhere.
+    ///
+    /// - Parameter bytes: Raw bytes received, ready to decode. On exit, any consumed bytes
+    ///                    will no longer be in the buffer.
+    /// - Returns: A decoded frame type or `nil` if more bytes are needed to complete the frame.
+    /// - Throws: Errors specified by RFC 7540 § 6 for mis-sized frames or incorrect
+    ///           stream/connection association.
+    func decode(bytes: inout ByteBuffer) throws -> HTTP2Frame? {
+        switch self.state {
+        case .idle:
+            return try readPotentialFrame(from: &bytes)
+        case .accumulatingData(let header, var currentBytes):
+            let required = header.length - currentBytes.readableBytes
+            if bytes.readableBytes >= required {
+                let view = bytes.viewBytes(at: bytes.readerIndex, length: required)
+                currentBytes.write(bytes: view)
+                bytes.moveReaderIndex(forwardBy: required)
+                return try readFrame(withHeader: header, from: &currentBytes)
+            } else {
+                currentBytes.write(buffer: &bytes)  // consume entire input buffer
+                return nil  // no frame (yet)
+            }
+        }
+    }
+    
+    private func readPotentialFrame(from bytes: inout ByteBuffer) throws -> HTTP2Frame? {
+        guard let header = bytes.readFrameHeader() else {
```

Comment:
Do we want to add `assert` statements to validate our assumptions about what state we're in?

Reply:
Sounds like a good idea.